wallet/walletrpc.c

```diff
@@ -177,75 +177,125 @@ static const struct json_command newaddr_command = {
+			 p_opt_def("start", param_u64, &liststart, 1),
+			 p_opt("limit", param_u32, &listlimit),
+			 p_opt_dev("dev_future", param_bool, &dev_future, false),
+			 NULL))
 		return command_param_failed();
 
-	if (!bip32_max_index) {
-		bip32_max_index = tal(cmd, u64);
-		*bip32_max_index = db_get_intvar(cmd->ld->wallet->db,
-						 "bip32_max_index", 0);
+	u64 bip32_max_index = db_get_intvar(cmd->ld->wallet->db, "bip32_max_index", 0);
+	addr = encode_scriptpubkey_to_addr(tmpctx, chainparams, scriptpubkey);
+
+	if (*liststart == 0) {
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						"Starting keyidx is 1; Cannot {start} with 0");
+	}
+	if (*dev_future && !listlimit) {
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						"Cannot use {dev_future} without {limit}");
+	}
+	struct issued_address_type *listaddrtypes = wallet_list_addresses(tmpctx, cmd->ld->wallet, *liststart, listlimit);
+	if (*dev_future) {
+		/* Limit can be greater than the issued addresses */
+		listend = *listlimit;
+	} else {
+		/* Maximum number of records returned from the query */
+		listend = tal_count(listaddrtypes);
 	}
 	response = json_stream_success(cmd);
 	json_array_start(response, "addresses");
-
-	for (s64 keyidx = 0; keyidx <= *bip32_max_index; keyidx++) {
-
-		if (keyidx == BIP32_INITIAL_HARDENED_CHILD){
+	for (size_t i = 0; i < listend; i++) {
+		curr_address = listaddrtypes[i];
+		/* Add future address types as `ADDR_ALL` if dev_future is true */
+		if (*liststart + i > bip32_max_index) {
+			curr_address.keyidx = *liststart + i;
+			curr_address.addrtype = ADDR_ALL;
+		}
+		if (curr_address.keyidx == BIP32_INITIAL_HARDENED_CHILD){
 			break;
 		}
-
-		bip32_pubkey(cmd->ld, &pubkey, keyidx);
-
-		// bech32 : p2wpkh
-		u8 *redeemscript_p2wpkh;
-		char *out_p2wpkh = encode_pubkey_to_addr(cmd,
-							 &pubkey,
-							 ADDR_BECH32,
-							 &redeemscript_p2wpkh);
-		if (!out_p2wpkh) {
-			abort();
+		bip32_pubkey(cmd->ld, &pubkey, curr_address.keyidx);
+		char *out_p2wpkh = "";
+		char *out_p2tr = "";
+		u8 *redeemscript_p2wpkh = tal_arr(cmd, u8, 0);
+		if (curr_address.addrtype == ADDR_BECH32 || curr_address.addrtype == ADDR_ALL) {
+			out_p2wpkh = encode_pubkey_to_addr(cmd,
+								&pubkey,
+								ADDR_BECH32,
+								&redeemscript_p2wpkh);
+			if (!out_p2wpkh) {
+				abort();
+			}
 		}
-
-		// p2tr
-		char *out_p2tr = encode_pubkey_to_addr(cmd,
-						       &pubkey,
-						       ADDR_P2TR,
-						       /* out_redeemscript */ NULL);
-		if (!out_p2tr) {
-			abort();
+		if (curr_address.addrtype == ADDR_P2TR || curr_address.addrtype == ADDR_ALL) {
+			out_p2tr = encode_pubkey_to_addr(cmd,
+								&pubkey,
+								ADDR_P2TR,
+								/* out_redeemscript */ NULL);
+			if (!out_p2tr) {
+				abort();
+			}
+		}
+		if (!addr || streq(addr, out_p2wpkh) || streq(addr, out_p2tr)) {
+			if (!cmd->ld->developer) {
+				json_add_address_details(response, curr_address.keyidx, NULL, NULL, out_p2wpkh, out_p2tr);
```

Comment:
Yech, please don't make schema --developer dependent!  Leave dev-listaddrs if you have to (as undocumented).

Reply:
Keeping the `dev-listaddrs` made the code way cleaner than before.